### PR TITLE
[editor] Convert from OpeningHours to TimeTableSet.

### PR DESCRIPTION
**Pull request by mgsergio**

### CODE REVIEW

**Comment by igrechuhin** [comment]
LGTM

File: 3party/opening_hours/opening_hours.cpp [hunk 1]
```diff
@@ -748,4 +749,40 @@ bool OpeningHours::IsValid() const
...
+{
+  return std::any_of(begin(m_rule), end(m_rule), [](RuleSequence const & seq)
+                     {
+                       return seq.HasMonths();
+                     });
+}
+
+bool OpeningHours::HasWeekSelector() const
+{
+  return std::any_of(begin(m_rule), end(m_rule), [](RuleSequence const & seq)
```
Comment (ygorshenin) [inline 3party/opening_hours/opening_hours.cpp]:
```c++
std::any_of(begin(m_rule), end(m_rule), mem_fun(&RuleSequence::HasWeeks));
```
?

File: 3party/opening_hours/opening_hours.hpp [hunk 2]
```diff
@@ -69,6 +69,16 @@ class HourMinutes
 HourMinutes operator-(HourMinutes const & hm);
 std::ostream & operator<<(std::ostream & ost, HourMinutes const & hm);
 
+bool operator<(HourMinutes const & a, HourMinutes const & b)
```
Comment (ygorshenin) [inline 3party/opening_hours/opening_hours.hpp]: We need to mark these operators as inline, right?

File: 3party/opening_hours/rules_evaluation.cpp [hunk 3]
```diff
@@ -196,6 +196,8 @@ std::tm MakeTimetuple(time_t const timestamp)
 
 namespace osmoh
 {
+using ::operator==;
```
Comment (ygorshenin) [inline 3party/opening_hours/rules_evaluation.cpp]: Why do we need this?

File: base/enumerate.hpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  iterator begin() { return {m_begin, m_countFrom}; }
+  iterator end() { return {m_end, {}}; }
+  original_iterator const m_begin;
+  original_iterator const m_end;
+  TCounter const m_countFrom;
+};
+
+template <typename TCollection, typename TCounter = size_t>
+auto enumerate(TCollection && collection, TCounter const counter = {})
```
Comment (ygorshenin) [inline base/enumerate.hpp]: s/enumerate/Enumerate/g

File: base/enumerate.hpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  using hint = double;
+  using type = typename TCollection::const_iterator;
+};
+
+template <typename TCollection>
+using const_respective_iterator_t =
+    typename const_respective_iterator<typename std::remove_reference<TCollection>::type>::type;
+}  // namespace details
+
+template <typename TCounter, typename TElement>
```
Comment (ygorshenin) [inline base/enumerate.hpp]: s/TCounter/TIndex/g

File: 3party/opening_hours/rules_evaluation.cpp (hunk 3 above)
Comment (mgsergio) [inline 3party/opening_hours/rules_evaluation.cpp]: Cause of ADL. it selects ```osmoh::operator<(HourMinutes const &, HourMinutes const &)``` instead.
And it results in compile error.

File: 3party/opening_hours/opening_hours.cpp (hunk 1 above)
Comment (mgsergio) [inline 3party/opening_hours/opening_hours.cpp]: Cool😎

File: 3party/opening_hours/opening_hours.hpp (hunk 2 above)
Comment (mgsergio) [inline 3party/opening_hours/opening_hours.hpp]: Right! Thanks.

File: base/enumerate.hpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+struct const_respective_iterator
+{
+  using hint = int;
+  using type = typename TCollection::iterator;
+};
+
+template <typename TCollection>
+struct const_respective_iterator<TCollection const>
+{
+  using hint = double;
```
Comment (ygorshenin) [inline base/enumerate.hpp]: Why do we need this? T and T const are different types, right?

File: base/enumerate.hpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+template <typename TCollection>
+using const_respective_iterator_t =
+    typename const_respective_iterator<typename std::remove_reference<TCollection>::type>::type;
+}  // namespace details
+
+template <typename TCounter, typename TElement>
+struct IndexedElement : std::pair<TCounter, TElement>
+{
+  using base = std::pair<TCounter, TElement>;
```
Comment (ygorshenin) [inline base/enumerate.hpp]: s/base/TBase/g

File: base/enumerate.hpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "std/iterator.hpp"
+#include "std/utility.hpp"
+
+namespace my
+{
+namespace details
+{
+template <typename TCollection>
+struct const_respective_iterator
```
Comment (ygorshenin) [inline base/enumerate.hpp]: Please, use CamelCase where it's possible.

File: 3party/opening_hours/rules_evaluation.cpp (hunk 3 above)
Comment (deathbaba) [inline 3party/opening_hours/rules_evaluation.cpp]: Should it be inserted as a comment into the code?

File: 3party/opening_hours/opening_hours.cpp (hunk 1 above)
Comment (mpimenov) [inline 3party/opening_hours/opening_hours.cpp]: mem_fun is deprecated, use mem_fn
http://en.cppreference.com/w/cpp/utility/functional/mem_fun

File: 3party/opening_hours/rules_evaluation.cpp (hunk 3 above)
Comment (mgsergio) [inline 3party/opening_hours/rules_evaluation.cpp]: Sure it should)

File: base/enumerate.hpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+#include "std/utility.hpp"
+
+namespace my
+{
+namespace details
+{
+template <typename TCollection>
+struct ConstRespectiveIterator
+{
+  using hint = int;
```
Comment (yunikkk) [inline base/enumerate.hpp]: Подсказка?)

Comment (deathbaba) [inline base/enumerate.hpp]: Зачем это вообще нужно? Чем int плох?

**Comment by ygorshenin** [comment]
LGTM

**Comment by yunikkk** [comment]
LGTM

**Pull request merged**